### PR TITLE
Pairwise union reach surface model

### PR DESCRIPTION
**Pull request by uakyol**

### CODE REVIEW

File: src/models/pairwise_union_reach_surface.py
Comment (matthewclegg) [inline src/models/pairwise_union_reach_surface.py:119]: Is alpha the same as b in the doc?

Comment (matthewclegg) [inline src/models/pairwise_union_reach_surface.py:119]: non-overlapped?

File: src/models/pairwise_union_reach_surface.py
Comment (matthewclegg) [inline src/models/pairwise_union_reach_surface.py:47]: Why not use the notation in the "parameter estimation" section of the Pairwise Union Overlap model doc?

File: src/models/tests/pairwise_union_reach_surface_test.py
Comment (matthewclegg) [inline src/models/tests/pairwise_union_reach_surface_test.py:74]: generate

File: src/models/tests/pairwise_union_reach_surface_test.py
Comment (matthewclegg) [inline src/models/tests/pairwise_union_reach_surface_test.py:87]: In general, it's not a good idea to include randomness in unit tests, because this can make it hard to reproduce the results.

File: src/models/tests/pairwise_union_reach_surface_test.py
Comment (matthewclegg) [inline src/models/tests/pairwise_union_reach_surface_test.py:24]: You could probably just use GoergModel rather than inventing a new class.

**Review by matthewclegg — CHANGES_REQUESTED** [review]

File: src/models/pairwise_union_reach_surface.py
Comment (uakyol) [reply src/models/pairwise_union_reach_surface.py:18]: thank you, I wasn't committing it apparently

File: src/models/pairwise_union_reach_surface.py
Comment (jiayu-google) [inline src/models/pairwise_union_reach_surface.py:79]: This method is also for defining constraints, so better to change the method name, or merge it with the previous method. Explicitly, the previous method gives inequality constraint Ga <= h, while this method gives equality constraint Aa = b (with b = 0 here).

File: src/models/pairwise_union_reach_surface.py
Comment (jiayu-google) [inline src/models/pairwise_union_reach_surface.py:70]: better to add docstrings

**Review by jiayu-google — CHANGES_REQUESTED** [review]
Thank you Ugur!

My main suggestion is to doc the algorithm description. 3 options off the top of my head:
(1) Describe everything in docstrings. Self-contained but hard to type some formulas. 
(2) Type the formulas (using the corresponding notations like G1) in a markdown, and include the markdown in this repo. Needs additional work.
(3) Type the formulas in LaTeX/doc and upload pdf to this repo.

I'd be happy to discuss and help!

File: src/models/tests/pairwise_union_reach_surface_test.py
Comment (uakyol) [reply src/models/tests/pairwise_union_reach_surface_test.py:24]: I took it from the reach_curve_test. If you don't mind I would like to keep this since it is vert intuitive?
But, I can change if you think there is value in using GoergModel model here.

File: src/models/tests/pairwise_union_reach_surface_test.py
Comment (uakyol) [reply src/models/tests/pairwise_union_reach_surface_test.py:90]: I used a random seed but didn't put impressions as an arg because it is simply generated in this for loop and fits with function name, do you think this works?

File: src/models/pairwise_union_reach_surface.py
Comment (uakyol) [reply src/models/pairwise_union_reach_surface.py:80]: ah thank you!

File: src/models/pairwise_union_reach_surface.py
Comment (uakyol) [reply src/models/pairwise_union_reach_surface.py]: yes, sorry I misnamed it. This is the equality constraint matrix, fixed.

File: src/models/pairwise_union_reach_surface.py
Comment (uakyol) [reply src/models/pairwise_union_reach_surface.py:156]: no Matthew, z is equivalent to B in the doc.
and d in the doc is ->    d = r_hat(true reach) - alpha

does this make sense? I also outlined how the qp matrices are created in the docstring. If you prefer, I can definitely change the name of the variables.

File: src/models/pairwise_union_reach_surface.py
Comment (jiayu-google) [inline src/models/pairwise_union_reach_surface.py:110]: Perhaps clarify the return: 
Returns (matrix G, vector h). Then, cvxopt.solvers recognizes the constraint Gx <= h, where x is the parameter vector.

File: src/models/pairwise_union_reach_surface.py
Comment (jiayu-google) [inline src/models/pairwise_union_reach_surface.py:127]: Perhaps clarify the return:
Returns (matrix A, vector b). Then, cvxopt.solvers recognizes the constraint Ax = b, where x is the parameter vector.

**Review by jiayu-google — APPROVED** [review]
Great, thank you Ugur! Left two minor comments.

File: src/models/pairwise_union_reach_surface.py
Comment (matthewclegg) [inline src/models/pairwise_union_reach_surface.py:127]: Just curious, where in this code are you enforcing the equality constraint that a_{i,j} = a_{j,i}?

File: src/models/pairwise_union_reach_surface.py
Comment (matthewclegg) [reply src/models/pairwise_union_reach_surface.py:110]: +1

File: src/models/tests/pairwise_union_reach_surface_test.py
Comment (matthewclegg) [inline src/models/tests/pairwise_union_reach_surface_test.py:84]: Does this matrix satisfy the constraint that a[i,j] = a[j,i]?

File: src/models/tests/pairwise_union_reach_surface_test.py
Comment (matthewclegg) [inline src/models/tests/pairwise_union_reach_surface_test.py:98]: When I run this unit test, I get some additional output, as follows:

```
     pcost       dcost       gap    pres   dres
 0: -3.4733e+09 -3.4733e+09  2e+01  2e+00  3e-10
 1: -3.4733e+09 -3.4733e+09  3e+00  2e-02  3e-12
 2: -3.4733e+09 -3.4733e+09  3e-02  2e-04  3e-14
 3: -3.4733e+09 -3.4733e+09  3e-04  2e-06  3e-16
 4: -3.4733e+09 -3.4733e+09  4e-06  2e-08  6e-17
Optimal solution found.
```

Can you turn this off?

**Review by matthewclegg — CHANGES_REQUESTED** [review]

File: src/models/pairwise_union_reach_surface.py
Comment (jiayu-google) [reply src/models/pairwise_union_reach_surface.py:127]: OK, this was not enforced in my colab. So Ugur, please add this constraint (by expanding matrix A).

File: src/models/tests/pairwise_union_reach_surface_test.py
Comment (jiayu-google) [reply src/models/tests/pairwise_union_reach_surface_test.py:84]: Ditto

File: src/models/tests/pairwise_union_reach_surface_test.py
Comment (jiayu-google) [reply src/models/tests/pairwise_union_reach_surface_test.py:98]: Try add `solvers.options['show_progress'] = False` in the beginning

File: src/models/pairwise_union_reach_surface.py
Comment (jiayu-google) [reply src/models/pairwise_union_reach_surface.py:30]: +1

File: src/models/pairwise_union_reach_surface.py
Comment (uakyol) [reply src/models/pairwise_union_reach_surface.py:71]: sorry, it seems I mistakenly copy/pasted from somewhere

File: src/models/pairwise_union_reach_surface.py
Comment (uakyol) [reply src/models/pairwise_union_reach_surface.py:53]: ah yes, thank you.

File: src/models/pairwise_union_reach_surface.py
Comment (uakyol) [reply src/models/pairwise_union_reach_surface.py:61]: thanks for catching this.

File: src/models/tests/pairwise_union_reach_surface_test.py
Comment (uakyol) [reply src/models/tests/pairwise_union_reach_surface_test.py:98]: thank you! I added at the beginning of the pairwise_union_reach_surface file not the test. What do you think? Do you prefer I add to test or is this good?

Comment (matthewclegg) [reply src/models/tests/pairwise_union_reach_surface_test.py:105]: It's better to have it in pairwise_union_reach_surface.py (as it is now).  Thanks!

File: src/models/tests/pairwise_union_reach_surface_test.py
Comment (matthewclegg) [inline src/models/tests/pairwise_union_reach_surface_test.py:105]: All of this looks good.  I have one additional request, if you don't mind.  Could you please add another two unit tests:

(a) Two reach curves, A and B, with no overlap.  If r_A(i) is the reach curve for A and r_B(i) is the reach curve for B, then the pairwise union model should satisfy r(i_A, i_B) = r_A(i_A) + r_B(i_B).

(b) Two reach curves, A and B, where the viewers in each curve are chosen uniformly at random from a universe of size N.  In this case, the pairwise union model should satisfy r(i_A, i_B) = r_A(i_A) + r_B(i_B) - r_A(i_A) * r_B(i_B) / N.

Thanks for coding all of this.  It looks great!

**Review by matthewclegg — CHANGES_REQUESTED** [review]

File: src/models/pairwise_union_reach_surface.py
Comment (uakyol) [reply src/models/pairwise_union_reach_surface.py]: closing this based on offline conversation

File: src/models/tests/pairwise_union_reach_surface_test.py
Comment (uakyol) [reply src/models/tests/pairwise_union_reach_surface_test.py]: thank you Matthew, I added those test cases too. Hope these make sense.

**Review by matthewclegg — APPROVED** [review]

File: src/models/pairwise_union_reach_surface.py
Comment (matthewclegg) [inline src/models/pairwise_union_reach_surface.py:167]: Ditto.

File: src/models/pairwise_union_reach_surface.py
Comment (matthewclegg) [inline src/models/pairwise_union_reach_surface.py:215]: Ditto.

**Review by matthewclegg — CHANGES_REQUESTED** [review]

File: src/models/pairwise_union_reach_surface.py
Comment (jiayu-google) [reply src/models/pairwise_union_reach_surface.py:86]: +1

Comment (uakyol) [reply src/models/pairwise_union_reach_surface.py:86]: thank you for finding this. Should be fixed now. I also pulled your changes and applied the fix version and confirmed the test you wrote for M3 worked. Hope it works when these are propagated up the chain too.

**Review by matthewclegg — APPROVED** [review]
Thanks, Ugur!

**Comment by AaronRanAn** [comment]
@uakyol thank you for putting this together! I start to go through the outstanding PRs this week (quite slowly due to the sheer volume). 

I reference the doc about the pairwise union model and believe this PR implements it correctly. However, I am not really familiar with how the code is set up and hope you can help provide a test case to let me test a simple case of Pairwise model run?  After that, I am comfortable we merge this one in. 

Thanks!

**Review by AaronRanAn — APPROVED** [review]
Feel free to move forward if not blocked by other reviewers.

**Comment by uakyol** [comment]
Thank you so much Aaron! I was thinking perhaps an explanation of the unit tests would be useful. The main one here is "test_by_impressions" in it
1) We generate simulated reach curves - one for each publisher - and with decaying universe sizes.
2) We generate a random matrix A - assuming the generating process for reach is actually a pairwise union model- we generate the model parameters
3) then, using these two, we can generate the reach points. 
4) We then fit the model - since the generating process is the same model, the fitted model should have a really small error rate.

What do you think?

**Comment by AaronRanAn** [comment]
Glad I can help @uakyol! Yeah, knowing the main unit test with the description make a lot more sense for me now. :) 

If you don't mind, we could add those in before you merge. Thanks again!

**Pull request merged**
